src/textual/css/parse.py

```diff
@@ -205,9 +204,102 @@ def parse_declarations(css: str, path: str) -> Styles:
     return styles_builder.styles
 
 
-def parse(css: str, path: str) -> Iterable[RuleSet]:
+def _is_whitespace(token: Token) -> bool:
+    return token.name == "whitespace"
+
+
+def _unresolved(
+    variable_name: str, location: tuple[int, int]
+) -> UnresolvedVariableError:
+    line_idx, col_idx = location
+    return UnresolvedVariableError(
+        f"reference to undefined variable '${variable_name}' at line {line_idx + 1}, column {col_idx + 1}."
+    )
 
-    tokens = iter(tokenize(css, path))
+
+def substitute_references(tokens: Iterator[Token]) -> Iterable[Token]:
+    """Replace variable references with values by substituting variable reference
+    tokens with the tokens representing their values.
+
+    Args:
+        tokens (Iterator[Token]): Iterator of Tokens which may contain tokens
+            with the name "variable_ref".
+
+    Returns:
+        Iterable[Token]: Yields Tokens such that any variable references (tokens where
+            token.name == "variable_ref") have been replaced with the tokens representing
+            the value. In other words, an Iterable of Tokens similar to the original input,
+            but with variables resolved. Substituted tokens will have their referenced_by
+            attribute populated with information about where the tokens are being substituted to.
+    """
+    variables: dict[str, list[Token]] = defaultdict(list)
+    while tokens:
+        token = next(tokens, None)
+        if token is None:
+            break
+        if token.name == "variable_name":
+            variable_name = token.value[1:-1]  # Trim the $ and the :, i.e. "$x:" -> "x"
+            yield token
+
+            # Store the tokens for any variable definitions, and substitute
+            # any variable references we encounter with them.
+            leading_whitespace = True
```

Comment:
I reckon this can be simplified to avoid that `leading_whitespace` variable.

Could you have a loop prior to this which first gets all whitespace from the iterator until it reaches one of the variable tokens.